pageserver/src/tenant/storage_layer.rs

```diff
@@ -83,6 +89,201 @@ pub enum ValueReconstructResult {
     Missing,
 }
 
+#[derive(Debug)]
+pub struct LayerAccessStats(Mutex<LayerAccessStatsInner>);
+
+#[derive(Debug, Default, Clone)]
+struct LayerAccessStatsInner {
+    first_access: Option<LayerAccessStatFullDetails>,
+    count_by_access_kind: EnumMap<LayerAccessKind, u64>,
+    task_kind_flag: EnumSet<TaskKind>,
+    last_accesses: HistoryBuffer<LayerAccessStatFullDetails, 16>,
+    last_residence_changes: HistoryBuffer<LayerResidenceStatus, 16>,
+}
+
+#[derive(Debug, Clone)]
+struct LayerAccessStatFullDetails {
+    when: SystemTime,
+    task_kind: TaskKind,
+    access_kind: LayerAccessKind,
+}
+
+#[derive(Debug, Clone)]
+pub enum LayerResidenceStatus {
+    Resident {
+        timestamp: SystemTime,
+        /// If `true`, then this resident status marks the birth time of the layer.
+        created: bool,
```

Comment:
consider having a `loaded` flag, for both `Resident` and `Evicted`. Or better even, an enum

```
struct LayerResidenceEvent {
  status: Resident | Evicted
  reason: Load | Created | Change 
}
```

Reply:
See https://github.com/neondatabase/neon/pull/3540#issuecomment-1416864942 on rationale for this.